### PR TITLE
feat: Add differentiable continuous-wave (pulsar) signals — `ripplegw.cw`

### PR DESCRIPTION
**Pull request by isaac-cf-wong**
## TL;DR

Adds a new signal class to ripple: **continuous-wave (CW) signals from spinning
neutron stars**, ported from LALPulsar to differentiable JAX and exposed through
the existing `Waveform` API (`{"p", "c"}` polarizations). Two LAL functions are
reproduced — `XLALSimulateExactPulsarSignal` (exact geometric reference) and
`XLALGeneratePulsarSignal` (full barycentering + heterodyne + binary orbit).
Everything is `jit`/`grad`-compatible and runs on CPU/GPU. **No new runtime
dependencies** (`jax`, `jaxtyping` only) and **no `lal`/`lalpulsar` dependency at
runtime** — the barycentering is reimplemented and reads a JPL ephemeris file
directly.

Validated against LALPulsar to its own numerical floor, including an
overlap-loss comparison against the compiled C functions.

## Why

ripple currently covers compact-binary-coalescence (CBC) waveforms. CW signals
are a distinct, long-lived signal class (targeted searches, parameter
estimation, mock-data challenges). A differentiable, hardware-accelerated CW
generator fits the same gradient-based inference niche ripple targets (e.g.
Jim), and there has not been a JAX-native, autodiff CW generator with
LAL-level barycentering. This PR provides one.

## What's added

New subpackage `src/ripplegw/cw/`:

| module | role |
|---|---|
| `ephemeris.py`     | parse LALPulsar JPL ephemeris files (gzipped ASCII) → numpy arrays |
| `time_utils.py`    | leap-second table + GMST/GAST (from `XLALBarycenterEarth`) |
| `detectors.py`     | H1/L1/V1 geocentric locations (from `LALDetectors.h`) |
| `earth.py`         | full Earth state: SSB position, sidereal time, precession/nutation, Einstein delay, Earth–Sun vector |
| `barycenter.py`    | detector SSB position + geometric Roemer delay; full `emit.deltaT` |
| `pulsar_signal.py` | `exact_pulsar_polarizations`, `generate_pulsar_polarizations`, `generate_binary_pulsar_polarizations` (+ a JAX Kepler solver) |
| `waveform.py`      | `Waveform` subclasses `ExactPulsarSignal`, `PulsarSignal`, `BinaryPulsarSignal` |

`src/ripplegw/constants.py` gains one constant (`AU = LAL_AU_SI`). Nothing
existing is modified.

### Usage

```python
import jax.numpy as jnp
from ripplegw.cw import PulsarSignal     # or ExactPulsarSignal / BinaryPulsarSignal

wf = PulsarSignal("H1", "earth00-40-DE405.dat.gz", "sun00-40-DE405.dat.gz",
                  start_gps=1_000_000_000, n_spindowns=1)
t = jnp.arange(0, 1800, 1 / 16)          # seconds since start_gps
out = wf(t, dict(alpha=1.3, delta=-0.5, f0=12.3, phi0=1.1,
                 aplus=1.0, across=0.64, f1=-1.1e-9))
hp, hc = out["p"], out["c"]
```

The call signature, return dict, and `parameter_names` property mirror the CBC
waveforms, so the same tooling (batching with `vmap`, `jit`, gradients) applies.

## Design decisions worth a reviewer's attention

- **`{"p","c"}` output (not detector strain).** LAL's functions return a single
  detector strain `h = F₊h₊ + F₎h₎`. We return the source polarizations
  `h₊ = A₊cosΨ`, `h₎ = A₎sinΨ` and leave the antenna-pattern projection (and
  thus the polarization angle ψ, arm geometry) to the caller — matching every
  other waveform in ripple. The `{p,c} ↔ (A₁…A₄)` reduction is derived in the
  code comments and checked numerically.
- **Self-contained barycentering.** Rather than depend on `lalpulsar` at runtime
  or accept precomputed detector states, the package reimplements
  `XLALBarycenterEarth`/`XLALBarycenter` in JAX and parses the JPL ephemeris
  itself. This keeps the dependency footprint at `jax`/`jaxtyping` and makes the
  whole signal differentiable end-to-end.
- **Per-sample phase, not LAL's coarse-grid interpolation.**
  `XLALGeneratePulsarSignal` builds the source signal on a coarse SSB grid and
  interpolates to detector samples (a speed optimization). We evaluate the phase
  directly at every sample. This is *more* accurate (no interpolation error) and
  cleanly differentiable; the cost is irrelevant on accelerators.
- **Int+frac GPS time.** Sample times are carried as integer second + fractional
  second to preserve ~ns precision through the ephemeris lookup and delay
  arithmetic (float64 GPS at `t≈10⁹` only resolves ~0.1 µs). This makes ripple
  more precise than LAL's own REAL8-GPS phase (see validation).
- **`numpy` is host-side only** (ephemeris file parsing); it is already a
  transitive dependency of `jax`. The differentiable path is pure `jax.numpy`.

## Validation

LAL's top-level functions take a `PulsarSignalParams` that swiglal does not wrap
(anonymous nested structs), so they cannot be called from Python. They are
validated two ways:

1. **Component anchors** against the SWIG-exposed routines they call internally:
   - geometric `n·rDetector` vs `XLALBarycenter`: **~1×10⁻¹³ s** (machine precision)
   - full `emit.deltaT` (Roemer + Earth-rotation w/ precession+nutation + Einstein − Shapiro) vs `XLALBarycenter`: **1.1×10⁻¹³ s**
   - binary source phase vs `XLALGenerateSpinOrbitCW`: machine precision where LAL solves Kepler tightly (see below)
2. **Comparison against the compiled top-level functions** via a small C harness
   (`tests/cross_validation/c_harness/`) that fills the struct in C and calls
   `XLALSimulateExactPulsarSignal` / `XLALGeneratePulsarSignal`, then computes
   the overlap loss between ripple's reconstructed strain and the LAL strain.

Figure of merit: the **overlap loss** `1 − ⟨h₁|h₂⟩/√(⟨h₁|h₁⟩⟨h₂|h₂⟩)` (same
metric and numerically-stable form as `test_lal_overlap.py`), reported as
`log10`.

### vs the compiled functions

| comparison | log10 overlap loss |
|---|---|
| `XLALSimulateExactPulsarSignal` | −12.5 |
| `XLALGeneratePulsarSignal` (no heterodyne) | −12.0 |
| `XLALGeneratePulsarSignal` (heterodyned) | −12.0 |

### Parameter sweep (400 random sets, H1, `f₀∈[10,500] Hz`)

| model | median log10(1−O) | max |
|---|---|---|
| exact (isolated)    | −9.2  | −7.9 |
| generate (isolated) | −11.3 | −10.5 |
| binary              | −5.5  | −4.8 |

cw_overlap_exact
cw_overlap_generate
cw_overlap_binary

- **exact** — ripple vs compiled `XLALSimulateExactPulsarSignal`
- **generate** — ripple vs compiled `XLALGeneratePulsarSignal` (isolated)
- **binary** — ripple vs compiled `XLALGeneratePulsarSignal` (binary orbit)

**Key point: the residual is LAL's reference precision, not ripple's** — each
floor is decomposed and anchored:

- *exact*: tracks the analytic LAL REAL8-GPS floor `(2π f₀·ulp(t))²/24` ∝ f₀²
  (overlaid in the figure); sky-position independent. ripple sits on/below it.
- *generate*: LAL's internal interpolation floor (shrinks with finer
  `sourceDeltaT`/`dtDelayBy2`/`dtPolBy2`); ripple is per-sample.
- *binary*: LAL `GenerateSpinOrbitCW`'s Kepler tolerance `dxMax = 0.01/(f₀·P)`
  (~0.01 rad). ripple's Kepler residual is exactly 0; at `f₀=1000 Hz` (LAL tight)
  the binary phase agrees to **log10(1−O) = −15.2**.

So across parameter space ripple reproduces the compiled LAL functions to LAL's
numerical floor and is intrinsically more accurate.

## Testing & CI

- **Unit tests** (`tests/unit/test_cw_*.py`) — no LAL/data dependency, run by
  default: ephemeris reader, leap seconds, detectors, analytic phase / heterodyne
  / binary→isolated reductions, Kepler residual, `jit` & `grad` (incl. the binary
  default-reference path).
- **Cross-validation** (`tests/cross_validation/test_cw_exact_pulsar.py`) — gated
  on `lalpulsar` + an ephemeris file (`RIPPLE_EARTH_EPHEMERIS`), skips cleanly
  otherwise, consistent with the existing `cross_validation` suite.
- **C harness** (`tests/cross_validation/c_harness/`) — manual gold-standard
  comparison against the compiled functions; `README.md` has the build/run
  recipe; the parameter-sweep results are cached so figures can be restyled
  without recomputation.
- `ruff` and `pyright` are clean.

## Dependencies & footprint

- No new runtime dependencies. `numpy` (host-side ephemeris parsing) is already
  a transitive `jax` dependency.
- Cross-validation needs `lalsuite` (already the repo's `cross-validation`
  extra) plus a JPL ephemeris file. **The pip `lalsuite` wheel does not bundle
  the ephemeris files** — obtain `earth*`/`sun*` from an installed lalsuite or
  the LALSuite repo; the reader handles the standard `earthNN-MM-DExxx.dat[.gz]`
  format.

## Known limitations / follow-ups

- **Scope**: isolated and elliptic-binary pulsars, with optional heterodyning;
  `{"p","c"}` output. The exact model neglects relativistic timing exactly as
  the LAL reference does; the full model includes Einstein/Shapiro.
- **Reference epochs** (`ref_time_ssb`, `tp_ssb`) are float64 absolute SSB
  seconds (LAL uses `LIGOTimeGPS` int+ns). Sample times keep int+frac precision;
  only these scalar epochs are float, which is a near-constant phase offset
  (overlap-insensitive) plus a negligible binary mean-anomaly shift. An int+frac
  epoch API is a clean follow-up if ns-level absolute-phase fidelity is desired.
- **Ephemeris range**: sample times outside the ephemeris span are clamped (JAX
  gather semantics) rather than raising as LAL does; callers must keep the
  observation within the file's coverage (standard files span 2000–2040).

## Notes for the maintainer

- Generated artifacts under `tests/cross_validation/c_harness/` are not tracked:
  the sweep cache `cw_sweep_results.csv`, `sweep_params.csv`, `*.bin`, and the
  compiled harness binaries via a local `.gitignore`; the figures (`figures/*.png`)
  via the repo's existing `tests/cross_validation/.gitignore` (`figures/`). All
  regenerate from the tracked sources with `make_figs.py` — the figures are
  attached to the PR rather than committed, matching the CBC cross-validation
  convention.
- The `cw` classes are exported from `ripplegw.cw` (not the top-level
  `ripplegw` namespace / `waveform_preset`), since their constructors require a
  detector + ephemeris + observation epoch and don't fit the `f_ref`-style CBC
  preset signature. Easy to surface them differently if you'd rather.

## Release Notes

* **New Features**
  * Added continuous-wave (pulsar) signal generation via `ExactPulsarSignal`, `PulsarSignal`, and `BinaryPulsarSignal`, returning `{"p", "c"}` on a detector-relative time axis.
  * Included built-in ground detector definitions (H1, L1, V1).
  * Added JPL ephemeris support for accurate barycentring, enabling full (isolated and binary) timing models plus optional heterodyning.
* **Documentation**
  * Expanded README and docs with continuous-wave waveform support, ephemeris requirements, and usage guidance.
* **Tests**
  * Added cross-validation and unit tests, including ephemeris parsing and JAX compatibility checks.

### CODE REVIEW

**Comment by isaac-cf-wong** [comment]
Is there an estimated timeline for merging this PR? If there is any additional information or justification needed, I’d be happy to provide it.

**Comment by thomasckng** [comment]
@isaac-cf-wong The main reason for the delay is that this PR reflects two design issues and I want to do a major restructuring of the codebase before merging.

One issue is how/where to place data files that the waveforms require. Another issue is the current top-level interface is expecting IMRPhenom-like waveforms, and ultimately we want to have all our waveforms to be accessed via the top-level API. Both of these issues are also relevant to #126.

The restructuring I mentioned is just a completely restructuring of top-level `ripple` interface, so that IMRPhenom waveforms can be modularised out. Maybe also some restructuring of all the tests. (The tests are terribly structured right now because I made some bad design decisions before.) This is independent of the internal restructuring of the IMPRhenom code, so I don't expect this restructuring to take a lot of time. Unfortunately, I couldn't find time to do it.

If you urgently need this to be on PyPI, another temporary solution we can go with is to make a pre-release version with what you have right now.

**Comment by isaac-cf-wong** [comment]
> @isaac-cf-wong The main reason for the delay is that this PR reflects two design issues and I want to do a major restructuring of the codebase before merging.
> 
> One issue is how/where to place data files that the waveforms require. Another issue is the current top-level interface is expecting IMRPhenom-like waveforms, and ultimately we want to have all our waveforms to be accessed via the top-level API. Both of these issues are also relevant to #126.
> 
> The restructuring I mentioned is just a completely restructuring of top-level `ripple` interface, so that IMRPhenom waveforms can be modularised out. Maybe also some restructuring of all the tests. (The tests are terribly structured right now because I made some bad design decisions before.) This is independent of the internal restructuring of the IMPRhenom code, so I don't expect this restructuring to take a lot of time. Unfortunately, I couldn't find time to do it.
> 
> If you urgently need this to be on PyPI, another temporary solution we can go with is to make a pre-release version with what you have right now.

I can investigate the requirements and any potential conflicts, then open a PR for the top-level changes. I’d just like to get continuous-wave support into gwmock soon.

**Comment by thomasckng** [comment]
@isaac-cf-wong That would be great. Thanks for the help! I had a few requirements in mind for the restructuring:

1. There should be a simple top-level interface where users don’t need to care about the underlying implementation. Ideally there should be a single entry point where they provide the necessary configuration (e.g. `f_ref`), and then simply use the returned waveform object via `__call__()`.
2. For waveforms that require external data (e.g. JAXNRSur and the CW waveforms), users shouldn’t need to manage those files manually. If the required files are reasonably small, I’m happy for them to be installed automatically with the PyPI package. If they’re too large, making those waveform modules available as optional installation extras also seems like a reasonable solution.
3. I also think this is a good opportunity to clean up the test structure. The current tests, especially the cross-validation tests, have become difficult to follow. It would be nice if the restructuring made them more modular and easier to extend when new waveform families are added in the future. However, this is independent of the two points above, so it can also be addressed in a separate PR.

**Comment by thomasckng** [comment]
@coderabbitai pause
